### ISSUE TITLE
Some SQL queries use NOW() which uses the mysql timezone (UTC) and not the shop timezone

### ISSUE THREAD

**Issue by matthieu-rolland** [PrestaShop/PrestaShop#30828]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

When inserting or updating objects in DB, the `date_add` et `date_upd` are filled using PHP and the shop's timezone.

But in a few cases, the SQL queries use the `NOW()` SQL directive, which uses the MySQL server's timezone (UTC).

### Expected behavior

We should always use the shop's timezone in PHP, and not `NOW()`, using `date('Y-m-d H:i:s')` instead.

### Steps to reproduce

1. Your shop timezone should be Europe/Paris
2. Your MySQL server timezone should be UTC (it's the default if you didn't change anything you're good)
3. Create a cart rule for a specific customer
4. Login in FO with this customer
5. go to my account > my reductions

the cart rule will only appear in one or two hours depending on summer/winter time.

### PrestaShop version(s) where the bug happened

develop

**Comment by Progi1984** [comment]
FYI @matthieu-rolland Some uses : 
![image](https://user-images.githubusercontent.com/1533248/212056515-f6aeb460-3ea9-4629-9903-91b9a7393ebe.png)

**Comment by sallemiines** [comment]
Hello @matthieu-rolland 

Thanx for reporting the issue.

Following your steps I got this result :

![Localization-•-develop](https://user-images.githubusercontent.com/31663369/212360553-a4e06605-364a-4d44-9652-dbf839049e2e.png)


![Cart-Rules-Edit-test-cart-•-develop](https://user-images.githubusercontent.com/31663369/212360648-036e1675-3cef-434c-ab4c-fb93e1c30aa1.png)

**In FO  > my account  > there's no carte rules appears**

![Capture-d’écran-du-2023-01-13-16-41-50-png-1328×583-](https://user-images.githubusercontent.com/31663369/212360674-8f2b801e-e093-4b4b-a6d3-cc9a3a6dfb78.png)

Did I miss smthg ? 

PLease check & feedback ! 
Thank you !

**Comment by Hlavtox** [comment]
I think that's an issue in your configuration. By default, mysql should be using server time and it's timezone.

**Comment by hibatallahAouadni** [comment]
Hello @matthieu-rolland 

Since we had no news from you for more than 20 days, I'll close this ticket to avoid cluttering up the backlog.
Please note that you can always create a new one if further information pops up.

Thanks!

**Comment by Flowster** [comment]
Hi @sallemiines
As I see it you encountered the bug during your test.
The fact that your front user can't see the cart rule you just created for him demonstrates that something is not working.

I just tested it again on a fresh PS 8.0 installation and can confirm the bug is still present.

**Comment by florine2623** [comment]
Hello @matthieu-rolland ,

I can't reproduce the issue on my side.

My time zone is well set :
<img width="762" alt="Screenshot 2023-03-31 at 15 20 00" src="https://user-images.githubusercontent.com/16019289/229131879-9bf98886-f87b-46ea-849d-5d202bbb982a.png">

I have created a cart rule that is directly applicable for the customer :
Created the 31/03/2023 at 15h24. `Valid from 31/03/2023 - 15h to 01/04/2023 - 15h.`
<img width="1184" alt="Screenshot 2023-03-31 at 15 25 53" src="https://user-images.githubusercontent.com/16019289/229132560-57c3f48e-aa00-4e30-a359-d8ffff68a841.png">

I log into my customer account, I see it : ✅ 
<img width="1141" alt="Screenshot 2023-03-31 at 15 22 28" src="https://user-images.githubusercontent.com/16019289/229132110-73080009-f9f8-4d9c-9784-f9958a1e2435.png">


Is there more details that I can apply to reproduce the issue ?
Thanks!

**Comment by Flowster** [comment]
Hi @florine2623 can you create a custom sql query in the SQL Manager?
`Advanced Parameters` > `Database` > `SQL Manager` > `+ Add new SQL query` :
* SQL query name : `Test`
* SQL query : `SELECT NOW() FROM ps_access;`

Play this query, if the time shown is the same time as your shop timezone then you will not encounter the problem because your server is running the same timezone as your shop.

-----

If this is the case put your shop on another timezone in the futur for exemple Brazil timezone instead of Paris.
In this configuration your server will be on Paris timezone and your shop on Brazil timezone.

**Comment by Hlavtox** [comment]
Database timezone should match php timezone, on properly configured server.

**Comment by Flowster** [comment]
@Hlavtox so you assume every shop owner should know how to configure their database timezone? And should have the possibility to do it? 😉
I'm not even speaking of a databases holding multiple shops with different timezones.

**Comment by sudo-do** [comment]
> Database timezone should match php timezone, on properly configured server.

@Hlavtox Someone in russia can rent hosting from usa

**Comment by Hlavtox** [comment]
@Flowster That's responsibility of the hosting administrator.

@sudo-do Yeah, and he should configure it properly.

**Comment by sudo-do** [comment]
MySQL inherits the system time from Windows/Linux

**Comment by sudo-do** [comment]
PS:shared hosting

**Comment by florine2623** [comment]
Hello @Flowster ,

Thanks for the tips !
I have indeed reproduce the issue when my timezone is different.

Although, when I change back my timezone to match the timezone from my SQL server, I still don't see the Discount in my customer account.

I have reproduces the issue with PrestaShop version `8.0.x`, `8.1.x`, `develop`, I'll add this to the backlog so it can be fixed.

Please be aware that some issues might take a very long time to be resolved. If this one is important to you and you cannot wait for it to be fixed on the project’s own time, we strongly suggest you consider [contacting a professional](https://www.prestashop-project.org/support/) to help you. 

If you fix the issue on your end, please [contribute it back to the project](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/). Remember that the more people contribute, the better PrestaShop becomes for everyone.

Thank you

**Comment by Hlavtox** [comment]
PHP, Apache and Mysql timezones NEED TO BE IN SYNC.

You want to tell me that using `NOW()` or `CURRENT_TIMESTAMP` have been wrong this whole time? I strongly disagree with what is discussed here, sorry. Just google for a while yourself.

**Comment by sudo-do** [comment]
@Hlavtox on shared hosting, it may not be possible to configure these settings, which can lead to issues with these functions.

**Comment by sudo-do** [comment]
AFAK, some companies do not care having their timezone settings in sync with customers, instead set their timezone to GMT 0 and inform their customers of this policy, like steam.

**Comment by Flowster** [comment]
@matthieu-rolland @florine2623 should we create different issues for each occurrence of `NOW()` in the code? It might be complicated to find an easily reproductible bug example for each one of those.

**Comment by 23b** [comment]
If the argument is that you can set the MySQL server timezone, why have a timezone setting inside Prestashop at all if you can simply change the PHP timezone setting on the server?
Also, what do you do if you have multiple shops for different countries on one server (our case)?

My suggested fix is this:

config/config.inc.php after `@date_default_timezone_set(Configuration::get('PS_TIMEZONE'));` (line 158) add this:

`Db::getInstance()->execute('SET @@time_zone = \''.(new DateTime())->format('P').'\'');`

**Comment by 23b** [comment]
short comment

**Comment by mattgoud** [comment]
A fix has been submitted for the core as well as all native modules affected by this issue:

**Core**
- PrestaShop/PrestaShop#41596 — `CartRule.php`, `ProductSale.php`, `Connection.php`, `Product.php`

**Native modules**
- PrestaShop/ps_googleanalytics#177
- PrestaShop/blockwishlist#451
- PrestaShop/ps_emailsubscription#127
- PrestaShop/ps_facetedsearch#1234
- PrestaShop/pagesnotfound#42
- PrestaShop/statssearch#30
- PrestaShop/statsbestproducts#28

**Comment by jf-viguier** [comment]
@mattgoud what about @23b suggestion ? https://github.com/PrestaShop/PrestaShop/issues/30828#issuecomment-2296880150

**Comment by mattgoud** [comment]
> [@mattgoud](https://github.com/mattgoud) what about [@23b](https://github.com/23b) suggestion ? [#30828 (comment)](https://github.com/PrestaShop/PrestaShop/issues/30828#issuecomment-2296880150)

@jf-viguier Indeed, that seems much more efficient than replacing the Now() calls on each module. I'll discuss it with the team tomorrow at the daily stand-up to see which solution we choose.

**Comment by mattgoud** [comment]
Hello @jf-viguier, so I created another PR ( #41628 ) inspired by @23b 's solution (connection-level approach) and targeting the develop branch (a more robust solution, better suited for a future `9.20-BETA` version).

Regarding the PR #41596  ("file-by-file approach"), I think we can keep it for an eventually `9.1.x patch` (I changed the target branch to 9.1 because it's a less robust but very low-risk solution).

For the PR inspired by @23b , tests have been added, and I've worked with Claude to ensure it also works for legacy AND Doctrine connections, and to address all the edge cases I had in mind (multistore/multi-server, etc. - see details in the PR's full description).

**Comment by mattgoud** [comment]
PR #41596 closed. The issue will only be addressed on the develop branch (possibly for version 9.2.0-BETA) and using the connection-level approach => PR #41628
